### DISCUSSION TITLE
SmartMeters support in Berry (cfr. SMI in script)

### DISCUSSION THREAD

**Discussion by FulvioSpelta** [arendst/Tasmota#20710] [Q&A] [answered]
Hi all, I've found the SmartMeters support in the doc:
[https://tasmota.github.io/docs/Smart-Meter-Interface/](https://tasmota.github.io/docs/Smart-Meter-Interface/)
but i see that you have to compile tasmota with script support.

I think that it could be interesting to have the same smartmeter support directly into Berry avoiding the adding of the script support to save resources.

Questions:
1) is there some kind of SmartMeter support in the actual version of Berry ? (or is there some similar function/workaround) ?
2) is there a plan to let it available in berry ?

Use case:
Collect energy data (Volt, Ampere from load and battery) from an industrial UPS via modbus/rtu; in modbus terms i've to read some holding registers and apply scale factors.
Currently i'm using multiple modbusSend scheduled using a rule and collect data from mqtt applying scale factors in a node.js script that is subscribed to the modbusReceived topic and runs on the db (influx) host.

Thanks a lot
f

edit: typo, added use case.

**Comment by s-hadinger** [comment]
No there are currently no plans. Anyways I don't have such devices, and the variety of devices looks huge. Calling for volunteers?

**Reply by gemu2015** [reply]
sml driver supports all kinds of meters and protocols including modbus serial or TCP master or slave.  see the docs for examples.
sml and berry can coexist without any problem in case you want to combine them.

**Reply by FulvioSpelta** [reply]
@gemu2015 yes I've read the docs and i like a much the flexibility to define registry, scale factors and so on.
My concern is about the use of resource due to script support activation on top (or in parallel) to berry and tasmota "core", and i'm "assuming" that a specific implementation in or for berry could be much more efficient overall (resource, load, ...).

Anyway have you kindly an example about how to define a SM in SMI and get the variables in berry ?
Is not clear to me if using tasmota.read_sensors() will return these variables.

**Reply by gemu2015** [reply, marked as answer]
simply take an example script from any modbus meter from the docs,  then type tasmota.read_sensors() in the berry console and you will get all the json values from the meter

**Reply by FulvioSpelta** [reply]
Thanks. I'll try

**Comment by FulvioSpelta** [comment]
@s-hadinger if you or someone else want to develop such a feature i'll be happy to collaborate on testing.
Feel free to contact me here or in PM.

**Comment by mi-hol** [comment]
> I think that it could be interesting to have the same smart meter support directly into Berry avoiding the adding of the script support

I have the very same requirement to avoid the cryptic definition of a smart meter in the old scripting language. 
The parallel use of old scripting and Berry causes unnecessary confusions leading to usability issues (example [here](https://github.com/arendst/Tasmota/discussions/22541#discussioncomment-11358294)).

Because reading the definitions made via old scripting is already working in Berry I wonder if there is already a way to define a smart meter via Berry that just needs to be documented?
